### PR TITLE
Satoshi dice

### PR DESCRIPTION
**Pull request by Eis-D-Z**
!DRAFT PR~
This PR will add a new section under examples for games/modules that have probabilities and chance. 

Only one module is being added for now, the satoshi flip where:
1) A player chooses a secret, commits on chain its hash and a Coin.
2) Another player ventures a guess on the last bit of the first byte of the secret and a wager of value up to previous player's Coin value or a possibly a predetermined amount.
3) Initial player commits secret on chain and the winner gets paid.

Things that will be improved before the PR is ready for merging:
1) the max_bet variable is overloaded, it should become a u64 and add a new variable with playerA's coin
2) needs additional tests, the two that exist check if the game rolls well. Edge cases and instances of wrong play also need testing.
3) Remove redundant assertions that are checked by the runtime (or are guaranteed in some way)
4) Polish code a bit by either compressing the logic or adding more helper functions (eg. separate check for who won)

### CODE REVIEW

**Comment by Eis-D-Z** [comment]
The max_bet variable issue has been fixed.

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+        assert!(max_bet >= min_bet, EMinBetTooHigh);
+        let bank_data = BankData {
+            bank_balance: coin::into_balance(banker_coin),
+            min_bet,
+            max_bet
+        };
+
+        let new_game = Game {
+            id: object::new(ctx),
+            epoch: tx_context::epoch(ctx),
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:91]: This is not ideal. We should allow for a period-range for player B to place a bet (i.e., 10 epochs after game creation), so bets stay open for a longer time than just 2 epochs. So ideally we need an extra property to define for how many epoch the game is live.
Then `EpochsCancelAfter` will be triggered by either 2 epochs after the expiration (ie 12 epochs in total) or 2 epochs after player's B bet (obviously if player B has an early bet, we don't necessarily need to wait for expiration).

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+        game_outcome.secret
+    }
+
+    public fun guess(game: &mut Game): u8 {
+        let game_outcome = option::borrow(&game.outcome);
+        game_outcome.guess
+    }
+
+    // this is invoked by playerB
+    public entry fun bet(game: &mut Game, guess: u8, stake_coin: Coin<SUI>, stake_amount: u64, ctx: &mut TxContext) {
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:139]: Iiuc this is not safe! We need to assert that players B bet happened on the allowable period, otherwise they can just bet right before expiration and player A doesn't have enough time to reveal. See this combines with my other message.

**Review by kchalkias — CHANGES_REQUESTED** [review]

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    public fun guess(game: &mut Game): u8 {
+        let game_outcome = option::borrow(&game.outcome);
+        game_outcome.guess
+    }
+
+    // this is invoked by player
+    public entry fun bet(game: &mut Game, guess: u8, stake_coin: Coin<SUI>, stake_amount: u64, ctx: &mut TxContext) {
+        let house_data = option::borrow(&game.house_data);
+        assert!(stake_amount <= house_data.max_bet, EStakeTooHigh);
+        assert!(stake_amount > house_data.min_bet, EStakeTooLow);
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:145]: `stake_amount >= house_data.min_bet`

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+    struct Outcome has store {
+        secret: vector<u8>,
+        guess: u8,
+        player_won: bool
+    }
+
+    struct Game has key {
+        id: UID,
+        epoch: u64,
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:71]: this is `bet_placed_epoch` let's rename for clarity

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        bet_data: Option<BetData>,
+        outcome: Option<Outcome>
+    }
+
+    // fun(ctions)!
+
+    // this is invoked by house
+    // each game will be a shared object so the player can cancel it in case the house refuses to end after EpochsCancelAfter epochs have passed
+    public entry fun start_game(hash: vector<u8>, house_coin: Coin<SUI>, max_bet: u64, min_bet: u64, ctx: &mut TxContext) {
+        assert!(max_bet >= min_bet, EMinBetTooHigh);
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:85]: add an `assert!(min_bet > 0, EZeroMinBet);` as well

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move (hunk 1 above)
Comment (kchalkias) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:95]: In theory we don't need to store the epoch at game's creation, but I decided that's fine for now.

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+        };
+        option::fill(&mut game.outcome, outcome);
+        let player = option::extract(&mut game.player);
+        pay_player(player, game.house, stake, house_balance, ctx);
+    }
+
+    // helper functions
+
+    fun pay_player(player: address, house: address, stake: Balance<SUI>, house_balance: Balance<SUI>, ctx: &mut TxContext) {
+        // if bet is less than max_bet, return the difference to player A after paying the wins
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:249]: as you renamed "Player A" to "house" return the difference to the **house**

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 7]
```diff
@@ -0,0 +1,264 @@
+module games_with_chance::satoshi_flip {
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:1]: Add license header (at the top of the file) and public docs for every function, i.e., see https://github.com/MystenLabs/sui/blob/main/crates/sui-framework/sources/safe.move

For public docs use ///

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move (hunk 3 above)
Comment (kchalkias) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:176]: is this resolved?

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+    /// Called by anyone after the required epochs have passed and house has not revealed the secret.
+    /// The house automatically loses.
+    /// Check EpochsCancelAfter for the number of epochs required to pass after `fun bet` has been called
+    public entry fun cancel_game(game: &mut Game, ctx: &mut TxContext) {
+        // this can't be called on an ended game
+        assert!(option::is_none<Outcome>(&game.outcome), EGameAlreadyEnded);
+        // a bet has to have been placed
+        assert!(option::is_some<BetData>(&game.bet_data), ECannotCancelBeforeBetting);
+        // this call can only be called 2 epochs after the bet has been placed
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:268]: This comment is wrong (and generally when we refer to 2 epochs in the comments). Ie now we set it to 5, so ideally just replace number `2` with `EpochsCancelAfter`

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+    /// Called by anyone after the required epochs have passed and house has not revealed the secret.
+    /// The house automatically loses.
+    /// Check EpochsCancelAfter for the number of epochs required to pass after `fun bet` has been called
+    public entry fun cancel_game(game: &mut Game, ctx: &mut TxContext) {
+        // this can't be called on an ended game
+        assert!(option::is_none<Outcome>(&game.outcome), EGameAlreadyEnded);
+        // a bet has to have been placed
+        assert!(option::is_some<BetData>(&game.bet_data), ECannotCancelBeforeBetting);
+        // this can only be called  `CancelEpochsAfter` epochs after the bet has been placed
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:269]: fix double space  in `called __ CancelEpochsAfter`

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+    /// The player can only be a single address for each game
+    ///
+    ///  @ownership: Shared
+    ///
+
+    struct Game has key {
+        id: UID,
+        epoch: u64,
+        hashed_secret: Sha3256Digest,
+        house: address ,
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:95]: remove space after address

File: sui_programmability/examples/games_with_chance/tests/test_satoshi_flip.move [hunk 11]
```diff
@@ -0,0 +1,743 @@
...
+        // player checks the game details and places a bet
+        test_scenario::next_tx(scenario, player);
+        {
+            let coinB = test_scenario::take_from_sender<Coin<SUI>>(scenario);
+            let game_val = test_scenario::take_shared<Game>(scenario);
+            let ctx = test_scenario::ctx(scenario);
+
+            // check is house address the correct one
+            assert!(satoshi_flip::house(&game_val) == @0xBAE, EWronghouse);
+            //check the minimum bet
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/tests/test_satoshi_flip.move:67]: space before check

File: sui_programmability/examples/games_with_chance/tests/test_satoshi_flip.move [hunk 12]
```diff
@@ -0,0 +1,743 @@
...
+            assert!(satoshi_flip::min_bet(&game_val) == 100, EWrongMinBet);
+            //check maximun bet
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/tests/test_satoshi_flip.move:69]: space before check

**Review by kchalkias — COMMENTED** [review]
minor typo comments

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+    }
+
+    /// get the minimum stake the house is willing to accept
+    public fun min_bet(game: &Game): u64 {
+        let house_data = option::borrow(&game.house_data);
+        house_data.min_bet
+    }
+    
+    /// On an ended game (that has a non null outcome value) get if player won
+    public fun is_player_winner(game: &mut Game): bool {
```
Comment (benr-ml) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:152]: does this have to be mut?
also relevant for the next getters below

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+            };
+            option::fill(&mut game.outcome, outcome);
+
+            // no bet placed, return the balance to house
+            let HouseData {house_balance, min_bet: _, max_bet: _} = option::extract(&mut game.house_data);
+            let house_coins = coin::from_balance(house_balance, ctx);
+            transfer::transfer(house_coins, game.house);
+        } else {
+            let hash = sha3_256(secret);
+            if (hash != digest::sha3_256_digest_to_bytes(&game.hashed_secret)) {
```
Comment (benr-ml) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:222]: the house has no incentive to send an invalid hash even if we don't punish anyone, no?
if so, we can just use assert here

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+                let player = option::extract(&mut game.player);
+                // TODO: Maybe add some punishment for the house
+                transfer::transfer(coin::from_balance(stake, ctx), player);
+                transfer::transfer(coin::from_balance(house_balance, ctx), game.house);
+                abort EHashAndSecretDontMatch
+            };
+            // extract balances and guess
+            let HouseData {house_balance, min_bet: _, max_bet: _} = option::extract(&mut game.house_data);
+            let BetData {stake, guess} = option::extract(&mut game.bet_data);
+            let first_byte = vector::borrow(&secret, 0);
```
Comment (benr-ml) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:234]: assert that secret is not empty (i think that sha3 would still work)

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+        option::fill(&mut game.bet_data, bet_data);
+        option::fill(&mut game.player, tx_context::sender(ctx));
+    }
+
+    /// Called by the house either after `fun bet` has been called by some player to end the game,
+    /// either before bet was called to "cancel" the game.
+    ///
+    /// If the house has "forgotten" the secret, then it can use a random secret and it will
+    /// cancel the game, as long as `fun bet` has not been called.
+    public entry fun end_game(game: &mut Game, secret: vector<u8>, ctx: &mut TxContext) {
```
Comment (benr-ml) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:202]: consider receiving it by value (ownership) and destructing it
also in cancel_game

Comment (Eis-D-Z) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:202]: I can do that with a shared object? Can I use object::delete(`game.id`) then?

Comment (benr-ml) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:202]: Right, I see your question on that (I wasn't aware of that limitation...)

**Comment by Eis-D-Z** [comment]
All suggested changes / fixes are ready. Should we change the epoch condition for cancelling to drand rounds?

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move (hunk 3 above)
Comment (Eis-D-Z) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move]: yes, now...

**Comment by kchalkias** [comment]
> All suggested changes / fixes are ready. Should we change the epoch condition for cancelling to drand rounds?

No, this version will deal with Epoch boundaries only. The mixed version will be a different example contract.

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+    */
+
+    /*
+        Conventions:
+        A guess = 2 in the outcome means player has not bet but the game was ended
+        A guess = 3 in the outcome means player has bet but house refused to reveal in time
+    */
+
+    /// Hash function the house will use
+    const HASH: vector<u8> = b"SHA3-256";
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:40]: Where is this used? if it's just a comment, just put it in a comment

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 18]
```diff
@@ -0,0 +1,300 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+
+/// Satoshi flip is a fair and secure way to conduct a coin flip.
+/// A player called "house" chooses a random secret and uploads its hash on chain
+/// Another player called "player",  makes a guess on a predetermined bit (eg. last) of the secret
+/// House reveals the secret and the winner is determined
+///
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:9]: I'd a comment about the optimistic version, let's call the game (in comments) "Optimistic Satoshi Dice" and explain that the main difference between the original Satoshi dice algorithm used in Bitcoin is that if the house refuses to reveal the secret on time, player can win the game.

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+    const EHouseCoinNotEnough: u64 = 9; // House provided coin with insuficient balance to cover max bet
+    const EPlayerCoinNotEnoughBalance: u64 = 10; // Player provided a coin with insufficient balance to cover his stake
+    const EGameNotEnded: u64 = 11;
+    const EAlreadyAcceptedBet: u64 = 12;
+    const ESecretIsEmpty: u64 = 13; // secret can't be an empty string
+    const EZeroMinBet: u64 = 14;
+
+    // structs
+
+    /// Player's data, the stake of the wager and his guess
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:65]: "his/her" or "the"

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+
+    /// Outcome will hold end game data, the secret the house picked, the player's guess and if the player won
+    /// This way anyone can check if the winner was correctly determined
+    struct Outcome has store {
+        secret: vector<u8>,
+        guess: u8,
+        player_won: bool
+    }
+
+    /// Each Game is a shared object, the address who created it must also end it.
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:87]: the address *that* created

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+        house_data: Option<HouseData>,
+        bet_data: Option<BetData>,
+        outcome: Option<Outcome>
+    }
+
+    // fun(ctions)!
+
+    /// creates a new game and makes it a transfered object
+    /// a user who wants to become house should call this function
+    public entry fun start_game(hash: vector<u8>, house_coin: Coin<SUI>, max_bet: u64, min_bet: u64, ctx: &mut TxContext) {
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:108]: I suggest ordering `min_bet` before `max_bet` in the inputs.

`min_bet: u64, max_bet: u64,`

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+        assert!(min_bet > 0, EZeroMinBet);
+        assert!(max_bet >= min_bet, EMinBetTooHigh);
+        assert!(coin::value(&house_coin) >= max_bet, EHouseCoinNotEnough);
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:111]: Add a comment here that we accept `house_coin_balance >= max_bet`, although in practice they should actually match. We allow this just because in Sui everything is an object (and thus to avoid splitting coins), but this contract's logic ensures that any surplus will return to the house, so it's safe to accept `house_coin_balance >= max_bet`.
The recommendation is (of course) for the house to pick a coin as close to max_bet as possible to avoid locking higher balances.

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+
+                balance::join(&mut house_balance, stake);
+                transfer::transfer(coin::from_balance(house_balance, ctx), game.house);
+            };
+        }
+    }
+
+    /// Called by anyone after the required epochs have passed and house has not revealed the secret.
+    /// The house automatically loses.
+    /// Check EpochsCancelAfter for the number of epochs required to pass after `fun bet` has been called
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:259]: generally prefer adding full stops to any comment (this applies to the whole PR)

File: sui_programmability/examples/games_with_chance/tests/test_satoshi_flip.move [hunk 24]
```diff
@@ -0,0 +1,724 @@
...
+    }
+
+    // Check that house provided coin of sufficient amount to cover the max_bet
+    #[test]
+    #[expected_failure(abort_code = 9)]
+    fun house_insufficient_balance() {
+        let world = @0x1EE7; // needed only for beginning the test_scenario
+        let house = @0xBAE;
+        let player = @0xFAB;
+        let secret = b"supersecret";
```
Comment (kchalkias) [inline sui_programmability/examples/games_with_chance/tests/test_satoshi_flip.move:306]: Note that ideally we want to recommend secrets being unpredictable random bytes of size at least 16 bytes. Please add this statement to the docs as well.

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 25]
```diff
@@ -0,0 +1,300 @@
...
+        A guess = 2 in the outcome means player has not bet but the game was ended
+        A guess = 3 in the outcome means player has bet but house refused to reveal in time
+    */
+
+    /// Hash function the house will use
+    const HASH: vector<u8> = b"SHA3-256";
+
+    /// How many epochs must pass after `fun bet` was called, until the player may cancel a game
+    // TODO: This should be possible to be set by house up to a max limit
+    const EpochsCancelAfter: u64 = 7;
```
Comment (benr-ml) [inline sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:44]: why 7? is 2 epochs enough? the user may send its bet just before the end of an epoch, and so we can't use here 1, but more than 1 should be safe

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move (hunk 17 above)
Comment (Eis-D-Z) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:40]: I put it in a constant that in order to be world readable (in explorer, with a call, by another module). I felt it was important since anyone trying to become house should be able to gain this info otherwise (if house chooses a random hash function) the house won't be able to end any games. It's a way of saying "It was written right there that you should use SHA3-256".
The practice is that such info is displayed in the comments?

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move (hunk 25 above)
Comment (Eis-D-Z) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:44]: This was requested by @kchalkias, it is like a standard in the gaming industry.

Comment (Eis-D-Z) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:44]: The actual reason is that in practice sometimes a random failure may happen, if it is the weekend and no person is on call the house might lose a winning game. One step further, along with the weekend, holidays might interimpose so the industry in general settled at 7 days. For our demo purposes it makes little difference, so we just went with the standard.

Comment (kchalkias) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move:50]: > why 7? is 2 epochs enough? the user may send its bet just before the end of an epoch, and so we can't use here 1, but more than 1 should be safe

7 days is the norm for fraud proof systems. We have to take into network outages, system/custody malfunction, lack of funds (so you need to find some to pay for gas), weekends and holidays (where services might be sub operating), potential access to cold keys (which requires some time due to custody deps). @Eis-D-Z please add a comment with the above and I believe we're good.

File: sui_programmability/examples/games_with_chance/sources/satoshi_flip.move [hunk 26]
```diff
@@ -0,0 +1,300 @@
...
+        assert!(option::is_none(&game.bet_data), EAlreadyAcceptedBet);
+        let house_data = option::borrow(&game.house_data);
+        assert!(stake_amount <= house_data.max_bet, EStakeTooHigh);
+        assert!(stake_amount >= house_data.min_bet, EStakeTooLow);
+        assert!(guess == 0 || guess == 1, EGuessNot1Or0);
+        assert!(coin::value(&stake_coin) >= stake_amount, EPlayerCoinNotEnoughBalance);
+        game.bet_placed_epoch = tx_context::epoch(ctx);
+        // Get a balance with the stake_amount value
+        let stake = coin::into_balance(stake_coin);
+        if (balance::value(&stake) > stake_amount) {
```
[the remark this answers is not in the document]
Comment (kchalkias) [reply sui_programmability/examples/games_with_chance/sources/satoshi_flip.move]: It doesn't matter re over optimizing for gas, cause the house can originally batch split for many independent games in one tx. I also like @benr-ml's comment re security, an average non-tech technical user wouldn't necessarily trust or can audit that contract; thus by just accepting a single coin as the max bet is totally fine.

**Review by kchalkias — COMMENTED** [review]
I think we're close to completion. Just add the comments to justify 7 days and the one coin that defines the max automatically (no change back to house). This will also simplify audits.

**Comment by Eis-D-Z** [comment]
All suggested changes have been committed.

**Comment by Eis-D-Z** [comment]
Contract will go under dapps with its UI logic.
